Digital 1/Assets/Scripts/ColoredObj.cs

```diff
@@ -0,0 +1,91 @@
+﻿using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+
+namespace Digital1
+{
+    /// <summary>
+	/// Base class for colored objects (i.e. player, bullets, enemies).
+	/// </summary>
+    public class ColoredObj : MonoBehaviour
+    {
+        /// <summary>
+        /// List of potential color states of the player/bullet/enemy.
+        /// </summary>
+        public enum ColorState
+        {
+            Neutral,
+            Red,
+            Blue,
+            Yellow
+        }
+
+        //[SerializeField]
+        private Material currentMaterial;
+        //[SerializeField]
+        private MeshRenderer meshRenderer;
+
+        [SerializeField]
+        protected ColorState currentState;
+        public ColorState CurrentState
+        {
+            get { return currentState; }
+        }
+
+        /// <summary>
+        /// Contains list of materials for color changing.
+        /// </summary>
+        [SerializeField]
+        private Material[] colorRef;
+
+        /// <summary>
+		/// Start is called before the first frame update
+		/// </summary>
+		/// <remarks>Known problem where meshRenderer is not initialized here; currently effective workaround is in <see cref="UpdateMaterial"/>.</remarks>
+        void Start()
+        {
+            currentState = ColorState.Neutral;
+            currentMaterial = GetComponent<Material>();
+            meshRenderer = GetComponent<MeshRenderer>();
+			if (meshRenderer == null)
+			{
+                meshRenderer = GetComponentInChildren<MeshRenderer>();
+            }
+        }
+
+        /// <summary>
+        /// Handle Color State switching, then update Material/Mesh.
+        /// </summary>
+        /// <param name="colorIndex"></param>
+        protected void ColorSwitch(int colorIndex)
+        {
+            switch (colorIndex)
+            {
+                case 1:
+                    currentState = ColorState.Red;
+                    break;
+                case 2:
+                    currentState = ColorState.Blue;
+                    break;
+                case 3:
+                    currentState = ColorState.Yellow;
+                    break;
+                default:
+                    currentState = ColorState.Neutral;
+                    break;
+            }
+
+            UpdateMaterial();
+        }
+
+        void UpdateMaterial()
+        {
+            currentMaterial = colorRef[(int)currentState];
```

Comment:
ColorState enum values should correlate directly to a number. Rather than assuming that the color starts at 0.

Reply:
Should also probably add a check to ensure that the index being grabbed exists.

Reply:
I changed the ColorState definition to manually define values, but I'm not sure if this is what you meant.
`public enum ColorState
    {
        Neutral = 0,
        Red     = 1,
        Blue    = 2,
        Yellow  = 3
    }`
I also added the check.
```
// currentMaterial = colorRef?[(int)currentState] ?? currentMaterial; // I think Unity has null operator problems, not sure. If not, this is better.
try
{
     currentMaterial = colorRef[(int)currentState];
}
catch (System.IndexOutOfRangeException)
{
     Debug.LogWarning("IndexOutOfRangeException in ColoredObj.UpdateMaterial(). Failed to change Material.");
}```

Reply:
Ya that's what I meant. It's just safer to know the value has been set in the enum rather than assuming that it'll work.